tools/crm_resource_runtime.c

```diff
@@ -1508,6 +1512,34 @@ cli_resource_execute(const char *rsc_id, const char *rsc_action, GHashTable *ove
 
     params = generate_resource_params(rsc, data_set);
 
+    /* add meta_timeout env needed by some resource agents */
+    child = first_named_child(rsc->xml, "operations");
+    for (child = first_named_child(child, XML_ATTR_OP);
+         child != NULL; child = crm_next_same_xml(child)) {
+        if (safe_str_eq(action, crm_element_value(child, XML_NVPAIR_ATTR_NAME))) {
+            timeout = crm_element_value(child, XML_ATTR_TIMEOUT);
+            break;
+        }
+    }
+    if (timeout == NULL && data_set->op_defaults) {
+        GHashTable *action_meta = crm_str_table_new();
+        unpack_instance_attributes(data_set->input, data_set->op_defaults, XML_TAG_META_SETS,
+                                   NULL, action_meta, NULL, FALSE, data_set->now);
+        timeout = g_hash_table_lookup(action_meta, XML_ATTR_TIMEOUT);
+    }
+    if (timeout == NULL && data_set->config_hash) {
+        timeout = pe_pref(data_set->config_hash, "default-action-timeout");
+    }
+    if (timeout == NULL) {
+        timeout = CRM_DEFAULT_OP_TIMEOUT_S;
+    }
+    timeout_ms = crm_get_msec(timeout);
+    if (timeout_ms < 0) {
+        timeout_ms = 0;
+    }
+    g_hash_table_insert(params, strdup("CRM_meta_timeout"),
+                        crm_strdup_printf("%ld", timeout_ms));
+
```

Comment:
I think the new "for" loop should be a function in libpe_status, like: pe_get_configured_timeout(rsc, action, data_set)